actor/src/main/scala/org/apache/pekko/io/dns/IdGenerator.scala

```diff
@@ -58,4 +56,37 @@ private[pekko] object IdGenerator {
    */
   def random(rand: java.util.Random): IdGenerator =
     () => (rand.nextInt(MaxUnsignedShort) - Short.MinValue).toShort
+
+  private[pekko] class EnhancedDoubleHashGenerator(semaphoreTimeout: FiniteDuration, seed: Random) extends IdGenerator {
+
+    /**
+     * An efficient thread safe generator of pseudo random shorts based on
+     * https://en.wikipedia.org/wiki/Double_hashing#Enhanced_double_hashing.
+     *
+     * Note that due to the usage of a semaphore this method is optimized
+     * for the happy case (i.e. least contention) on multiple threads.
+     */
+    private val lock = new Semaphore(1)
+    private var index = seed.nextLong
+    private var increment = seed.nextLong
+    private var count = 1L
+
+    private val onAcquireFailure = IdGenerator.random(seed)
+
+    override final def nextId(): Short = {
+      if (lock.tryAcquire(semaphoreTimeout.length, semaphoreTimeout.unit)) {
```

Comment:
revert to synchronized if you find it is working better for you

Reply:
Okay I will do so

Reply:
btw I just merged a PR fixing a small issue in the negative ID fix - I have resolved the conflict in your PR/branch - so you might want to pull the latest code